src/sentry/services/hybrid_cloud/rpc.py

```diff
@@ -162,16 +163,33 @@ def deserialize_return_value(self, value: Any) -> Any:
         parsed = self._return_model.parse_obj({self._RETURN_MODEL_ATTR: value})
         return getattr(parsed, self._RETURN_MODEL_ATTR)
 
-    def resolve_to_region(self, arguments: ArgumentDict) -> Region:
+    def resolve_to_region(self, arguments: ArgumentDict) -> _RegionResolutionResult:
         if self._region_resolution is None:
             raise RpcServiceSetupException(f"{self.service_name} does not run on the region silo")
 
         try:
-            return self._region_resolution.resolve(arguments)
+            try:
+                region = self._region_resolution.resolve(arguments)
+                return _RegionResolutionResult(region)
+            except RegionMappingNotFound:
+                if getattr(self._base_method, _REGION_RESOLUTION_OPTIONAL_RETURN_ATTR, False):
+                    return _RegionResolutionResult(None, is_early_halt=True)
+                else:
+                    raise
         except Exception as e:
             raise RpcServiceUnimplementedException("Error while resolving region") from e
 
 
+@dataclass(frozen=True)
+class _RegionResolutionResult:
+    region: Region | None
+    is_early_halt: bool = False
```

Comment:
See my commit log on https://github.com/getsentry/sentry/pull/52560/commits/236d47adf7ca4fd75186395d961cbccf38f55ff7 for rationale.